curl@7.84.0 on ppc64le: easy_lock.h:56:7: error: implicit declaration of function 'sched_yield' [-Werror=implicit-function-declaration]

### Steps to reproduce the issue

`curl@7.84.0` build fails on ppc64le due to:
* https://github.com/curl/curl/issues/9136

This was fixed in curl master:
* https://github.com/curl/curl/pull/9054

Failure reproducible on:
* Ubuntu 18.04, GCC 7.5.0, ppc64le
* Ubuntu 20.04, GCC 11.1.0, ppc64le
* Ubuntu 22.04, GCC 11.2.0, ppc64le

Reproduced below using Docker container image:
* `ecpe4s/ubuntu20.04-runner-ppc64le:2022-07-01`
```
$> git clone https://github.com/spack/spack
$> (cd spack && git checkout 519a2d2b154263a14c8c24a9d7f12167bad7295e)
$> . spack/share/spack/setup-env.sh

$> spack install curl@7.84.0
...
==> Installing curl-7.84.0-bj443wb66znlrie5nwn5byvcbzcotree
==> Fetching http://curl.haxx.se/download/curl-7.84.0.tar.bz2
==> No patches needed for curl
==> curl: Executing phase: 'autoreconf'
==> curl: Executing phase: 'configure'
==> curl: Executing phase: 'build'
==> Error: ProcessError: Command exited with status 2:
    'make' '-j16' 'V=1'

4 errors found in build log:
     834    /bin/bash ../libtool  --tag=CC   --mode=compile /spack/lib/spack/env/gcc/gcc -DHAVE_CONFIG_H   -I../include -I../lib -I../lib -DBUILDING_LIBCURL   -isystem /spack/o
            pt/spack/linux-ubuntu20.04-power9le/gcc-11.1.0/zlib-1.2.12-iwxgndy2ndnrpyhz44hwzaarvowhiryf/include -isystem /spack/opt/spack/linux-ubuntu20.04-power9le/gcc-11.1.0/
            openssl-1.1.1q-nrnwthzzlsb4hcvmfmspgm42ry3wye6y/include   -Werror-implicit-function-declaration -O2 -Wno-system-headers -pthread  -MT libcurl_la-escape.lo -MD -MP -
            MF .deps/libcurl_la-escape.Tpo -c -o libcurl_la-escape.lo `test -f 'escape.c' || echo './'`escape.c
     835    libtool: compile:  /spack/lib/spack/env/gcc/gcc -DHAVE_CONFIG_H -I../include -I../lib -I../lib -DBUILDING_LIBCURL -isystem /spack/opt/spack/linux-ubuntu20.04-power9
            le/gcc-11.1.0/zlib-1.2.12-iwxgndy2ndnrpyhz44hwzaarvowhiryf/include -isystem /spack/opt/spack/linux-ubuntu20.04-power9le/gcc-11.1.0/openssl-1.1.1q-nrnwthzzlsb4hcvmfm
            spgm42ry3wye6y/include -Werror-implicit-function-declaration -O2 -Wno-system-headers -pthread -MT libcurl_la-easygetopt.lo -MD -MP -MF .deps/libcurl_la-easygetopt.T
            po -c easygetopt.c  -fPIC -DPIC -o .libs/libcurl_la-easygetopt.o
     836    libtool: compile:  /spack/lib/spack/env/gcc/gcc -DHAVE_CONFIG_H -I../include -I../lib -I../lib -DBUILDING_LIBCURL -isystem /spack/opt/spack/linux-ubuntu20.04-power9
            le/gcc-11.1.0/zlib-1.2.12-iwxgndy2ndnrpyhz44hwzaarvowhiryf/include -isystem /spack/opt/spack/linux-ubuntu20.04-power9le/gcc-11.1.0/openssl-1.1.1q-nrnwthzzlsb4hcvmfm
            spgm42ry3wye6y/include -Werror-implicit-function-declaration -O2 -Wno-system-headers -pthread -MT libcurl_la-dotdot.lo -MD -MP -MF .deps/libcurl_la-dotdot.Tpo -c do
            tdot.c -o libcurl_la-dotdot.o >/dev/null 2>&1
     837    mv -f .deps/libcurl_la-curl_sspi.Tpo .deps/libcurl_la-curl_sspi.Plo
     838    In file included from easy.c:89:
     839    easy_lock.h: In function 'curl_simple_lock_lock':
  >> 840    easy_lock.h:56:7: error: implicit declaration of function 'sched_yield' [-Werror=implicit-function-declaration]
     841       56 |       sched_yield();
     842          |       ^~~~~~~~~~~
     843    /bin/bash ../libtool  --tag=CC   --mode=compile /spack/lib/spack/env/gcc/gcc -DHAVE_CONFIG_H   -I../include -I../lib -I../lib -DBUILDING_LIBCURL   -isystem /spack/o
            pt/spack/linux-ubuntu20.04-power9le/gcc-11.1.0/zlib-1.2.12-iwxgndy2ndnrpyhz44hwzaarvowhiryf/include -isystem /spack/opt/spack/linux-ubuntu20.04-power9le/gcc-11.1.0/
            openssl-1.1.1q-nrnwthzzlsb4hcvmfmspgm42ry3wye6y/include   -Werror-implicit-function-declaration -O2 -Wno-system-headers -pthread  -MT libcurl_la-file.lo -MD -MP -MF
             .deps/libcurl_la-file.Tpo -c -o libcurl_la-file.lo `test -f 'file.c' || echo './'`file.c
     844    libtool: compile:  /spack/lib/spack/env/gcc/gcc -DHAVE_CONFIG_H -I../include -I../lib -I../lib -DBUILDING_LIBCURL -isystem /spack/opt/spack/linux-ubuntu20.04-power9
            le/gcc-11.1.0/zlib-1.2.12-iwxgndy2ndnrpyhz44hwzaarvowhiryf/include -isystem /spack/opt/spack/linux-ubuntu20.04-power9le/gcc-11.1.0/openssl-1.1.1q-nrnwthzzlsb4hcvmfm
            spgm42ry3wye6y/include -Werror-implicit-function-declaration -O2 -Wno-system-headers -pthread -MT libcurl_la-easyoptions.lo -MD -MP -MF .deps/libcurl_la-easyoptions
            .Tpo -c easyoptions.c  -fPIC -DPIC -o .libs/libcurl_la-easyoptions.o
     845    libtool: compile:  /spack/lib/spack/env/gcc/gcc -DHAVE_CONFIG_H -I../include -I../lib -I../lib -DBUILDING_LIBCURL -isystem /spack/opt/spack/linux-ubuntu20.04-power9
            le/gcc-11.1.0/zlib-1.2.12-iwxgndy2ndnrpyhz44hwzaarvowhiryf/include -isystem /spack/opt/spack/linux-ubuntu20.04-power9le/gcc-11.1.0/openssl-1.1.1q-nrnwthzzlsb4hcvmfm
            spgm42ry3wye6y/include -Werror-implicit-function-declaration -O2 -Wno-system-headers -pthread -MT libcurl_la-dict.lo -MD -MP -MF .deps/libcurl_la-dict.Tpo -c dict.c
             -o libcurl_la-dict.o >/dev/null 2>&1
     846    libtool: compile:  /spack/lib/spack/env/gcc/gcc -DHAVE_CONFIG_H -I../include -I../lib -I../lib -DBUILDING_LIBCURL -isystem /spack/opt/spack/linux-ubuntu20.04-power9
            le/gcc-11.1.0/zlib-1.2.12-iwxgndy2ndnrpyhz44hwzaarvowhiryf/include -isystem /spack/opt/spack/linux-ubuntu20.04-power9le/gcc-11.1.0/openssl-1.1.1q-nrnwthzzlsb4hcvmfm
            spgm42ry3wye6y/include -Werror-implicit-function-declaration -O2 -Wno-system-headers -pthread -MT libcurl_la-dynbuf.lo -MD -MP -MF .deps/libcurl_la-dynbuf.Tpo -c dy
            nbuf.c -o libcurl_la-dynbuf.o >/dev/null 2>&1

     ...

     875    libtool: compile:  /spack/lib/spack/env/gcc/gcc -DHAVE_CONFIG_H -I../include -I../lib -I../lib -DBUILDING_LIBCURL -isystem /spack/opt/spack/linux-ubuntu20.04-power9
            le/gcc-11.1.0/zlib-1.2.12-iwxgndy2ndnrpyhz44hwzaarvowhiryf/include -isystem /spack/opt/spack/linux-ubuntu20.04-power9le/gcc-11.1.0/openssl-1.1.1q-nrnwthzzlsb4hcvmfm
            spgm42ry3wye6y/include -Werror-implicit-function-declaration -O2 -Wno-system-headers -pthread -MT libcurl_la-ftplistparser.lo -MD -MP -MF .deps/libcurl_la-ftplistpa
            rser.Tpo -c ftplistparser.c  -fPIC -DPIC -o .libs/libcurl_la-ftplistparser.o
     876    mv -f .deps/libcurl_la-curl_ntlm_wb.Tpo .deps/libcurl_la-curl_ntlm_wb.Plo
     877    /bin/bash ../libtool  --tag=CC   --mode=compile /spack/lib/spack/env/gcc/gcc -DHAVE_CONFIG_H   -I../include -I../lib -I../lib -DBUILDING_LIBCURL   -isystem /spack/o
            pt/spack/linux-ubuntu20.04-power9le/gcc-11.1.0/zlib-1.2.12-iwxgndy2ndnrpyhz44hwzaarvowhiryf/include -isystem /spack/opt/spack/linux-ubuntu20.04-power9le/gcc-11.1.0/
            openssl-1.1.1q-nrnwthzzlsb4hcvmfmspgm42ry3wye6y/include   -Werror-implicit-function-declaration -O2 -Wno-system-headers -pthread  -MT libcurl_la-h2h3.lo -MD -MP -MF
             .deps/libcurl_la-h2h3.Tpo -c -o libcurl_la-h2h3.lo `test -f 'h2h3.c' || echo './'`h2h3.c
     878    libtool: compile:  /spack/lib/spack/env/gcc/gcc -DHAVE_CONFIG_H -I../include -I../lib -I../lib -DBUILDING_LIBCURL -isystem /spack/opt/spack/linux-ubuntu20.04-power9
            le/gcc-11.1.0/zlib-1.2.12-iwxgndy2ndnrpyhz44hwzaarvowhiryf/include -isystem /spack/opt/spack/linux-ubuntu20.04-power9le/gcc-11.1.0/openssl-1.1.1q-nrnwthzzlsb4hcvmfm
            spgm42ry3wye6y/include -Werror-implicit-function-declaration -O2 -Wno-system-headers -pthread -MT libcurl_la-getenv.lo -MD -MP -MF .deps/libcurl_la-getenv.Tpo -c ge
            tenv.c  -fPIC -DPIC -o .libs/libcurl_la-getenv.o
     879    libtool: compile:  /spack/lib/spack/env/gcc/gcc -DHAVE_CONFIG_H -I../include -I../lib -I../lib -DBUILDING_LIBCURL -isystem /spack/opt/spack/linux-ubuntu20.04-power9
            le/gcc-11.1.0/zlib-1.2.12-iwxgndy2ndnrpyhz44hwzaarvowhiryf/include -isystem /spack/opt/spack/linux-ubuntu20.04-power9le/gcc-11.1.0/openssl-1.1.1q-nrnwthzzlsb4hcvmfm
            spgm42ry3wye6y/include -Werror-implicit-function-declaration -O2 -Wno-system-headers -pthread -MT libcurl_la-file.lo -MD -MP -MF .deps/libcurl_la-file.Tpo -c file.c
             -o libcurl_la-file.o >/dev/null 2>&1
     880    cc1: some warnings being treated as errors
  >> 881    make[2]: *** [Makefile:2230: libcurl_la-easy.lo] Error 1
     882    make[2]: *** Waiting for unfinished jobs....
     883    libtool: compile:  /spack/lib/spack/env/gcc/gcc -DHAVE_CONFIG_H -I../include -I../lib -I../lib -DBUILDING_LIBCURL -isystem /spack/opt/spack/linux-ubuntu20.04-power9
            le/gcc-11.1.0/zlib-1.2.12-iwxgndy2ndnrpyhz44hwzaarvowhiryf/include -isystem /spack/opt/spack/linux-ubuntu20.04-power9le/gcc-11.1.0/openssl-1.1.1q-nrnwthzzlsb4hcvmfm
            spgm42ry3wye6y/include -Werror-implicit-function-declaration -O2 -Wno-system-headers -pthread -MT libcurl_la-fopen.lo -MD -MP -MF .deps/libcurl_la-fopen.Tpo -c fope
            n.c -o libcurl_la-fopen.o >/dev/null 2>&1
     884    libtool: compile:  /spack/lib/spack/env/gcc/gcc -DHAVE_CONFIG_H -I../include -I../lib -I../lib -DBUILDING_LIBCURL -isystem /spack/opt/spack/linux-ubuntu20.04-power9
            le/gcc-11.1.0/zlib-1.2.12-iwxgndy2ndnrpyhz44hwzaarvowhiryf/include -isystem /spack/opt/spack/linux-ubuntu20.04-power9le/gcc-11.1.0/openssl-1.1.1q-nrnwthzzlsb4hcvmfm
            spgm42ry3wye6y/include -Werror-implicit-function-declaration -O2 -Wno-system-headers -pthread -MT libcurl_la-gopher.lo -MD -MP -MF .deps/libcurl_la-gopher.Tpo -c go
            pher.c  -fPIC -DPIC -o .libs/libcurl_la-gopher.o
     885    libtool: compile:  /spack/lib/spack/env/gcc/gcc -DHAVE_CONFIG_H -I../include -I../lib -I../lib -DBUILDING_LIBCURL -isystem /spack/opt/spack/linux-ubuntu20.04-power9
            le/gcc-11.1.0/zlib-1.2.12-iwxgndy2ndnrpyhz44hwzaarvowhiryf/include -isystem /spack/opt/spack/linux-ubuntu20.04-power9le/gcc-11.1.0/openssl-1.1.1q-nrnwthzzlsb4hcvmfm
            spgm42ry3wye6y/include -Werror-implicit-function-declaration -O2 -Wno-system-headers -pthread -MT libcurl_la-h2h3.lo -MD -MP -MF .deps/libcurl_la-h2h3.Tpo -c h2h3.c
              -fPIC -DPIC -o .libs/libcurl_la-h2h3.o
     886    libtool: compile:  /spack/lib/spack/env/gcc/gcc -DHAVE_CONFIG_H -I../include -I../lib -I../lib -DBUILDING_LIBCURL -isystem /spack/opt/spack/linux-ubuntu20.04-power9
            le/gcc-11.1.0/zlib-1.2.12-iwxgndy2ndnrpyhz44hwzaarvowhiryf/include -isystem /spack/opt/spack/linux-ubuntu20.04-power9le/gcc-11.1.0/openssl-1.1.1q-nrnwthzzlsb4hcvmfm
            spgm42ry3wye6y/include -Werror-implicit-function-declaration -O2 -Wno-system-headers -pthread -MT libcurl_la-getenv.lo -MD -MP -MF .deps/libcurl_la-getenv.Tpo -c ge
            tenv.c -o libcurl_la-getenv.o >/dev/null 2>&1
     887    mv -f .deps/libcurl_la-connect.Tpo .deps/libcurl_la-connect.Plo

     ...

     904    mv -f .deps/libcurl_la-formdata.Tpo .deps/libcurl_la-formdata.Plo
     905    mv -f .deps/libcurl_la-doh.Tpo .deps/libcurl_la-doh.Plo
     906    mv -f .deps/libcurl_la-ftplistparser.Tpo .deps/libcurl_la-ftplistparser.Plo
     907    libtool: compile:  /spack/lib/spack/env/gcc/gcc -DHAVE_CONFIG_H -I../include -I../lib -I../lib -DBUILDING_LIBCURL -isystem /spack/opt/spack/linux-ubuntu20.04-power9
            le/gcc-11.1.0/zlib-1.2.12-iwxgndy2ndnrpyhz44hwzaarvowhiryf/include -isystem /spack/opt/spack/linux-ubuntu20.04-power9le/gcc-11.1.0/openssl-1.1.1q-nrnwthzzlsb4hcvmfm
            spgm42ry3wye6y/include -Werror-implicit-function-declaration -O2 -Wno-system-headers -pthread -MT libcurl_la-ftp.lo -MD -MP -MF .deps/libcurl_la-ftp.Tpo -c ftp.c -o
             libcurl_la-ftp.o >/dev/null 2>&1
     908    mv -f .deps/libcurl_la-ftp.Tpo .deps/libcurl_la-ftp.Plo
     909    make[2]: Leaving directory '/tmp/root/spack-stage/spack-stage-curl-7.84.0-bj443wb66znlrie5nwn5byvcbzcotree/spack-src/lib'
  >> 910    make[1]: *** [Makefile:1365: all] Error 2
     911    make[1]: Leaving directory '/tmp/root/spack-stage/spack-stage-curl-7.84.0-bj443wb66znlrie5nwn5byvcbzcotree/spack-src/lib'
  >> 912    make: *** [Makefile:1240: all-recursive] Error 1
```

### Error message

<details><summary>Error message</summary><pre>
... see above
</pre></details>


### Information on your system

* **Spack:** 0.19.0.dev0 (519a2d2b154263a14c8c24a9d7f12167bad7295e)
* **Python:** 3.8.10
* **Platform:** linux-ubuntu20.04-power9le
* **Concretizer:** clingo

### Additional information

[spack-build-out.txt](https://github.com/spack/spack/files/9269859/spack-build-out.txt)

### General information

- [X] I have run `spack debug report` and reported the version of Spack/Python/Platform
- [X] I have run `spack maintainers <name-of-the-package>` and **@mentioned** any maintainers
- [X] I have uploaded the build log and environment files
- [X] I have searched the issues of this repo and believe this is not a duplicate

--- a/var/spack/repos/builtin/packages/curl/easy-lock-sched-header.patch
+++ b/var/spack/repos/builtin/packages/curl/easy-lock-sched-header.patch
@@ -1,0 +1,13 @@
+diff -ruN spack-src/lib/easy_lock.h spack-src-patched/lib/easy_lock.h
+--- spack-src/lib/easy_lock.h	2022-06-26 13:58:05.000000000 -0700
++++ spack-src-patched/lib/easy_lock.h	2022-08-05 07:59:16.450137961 -0700
+@@ -36,6 +36,9 @@
+ 
+ #elif defined (HAVE_ATOMIC)
+ #include <stdatomic.h>
++#if defined(HAVE_SCHED_YIELD)
++#include <sched.h>
++#endif
+ 
+ #define curl_simple_lock atomic_bool
+ #define CURL_SIMPLE_LOCK_INIT false

--- a/var/spack/repos/builtin/packages/curl/easy-lock-sched-header.patch
+++ b/var/spack/repos/builtin/packages/curl/easy-lock-sched-header.patch
@@ -1,6 +1,19 @@
-diff -ruN spack-src/lib/easy_lock.h spack-src-patched/lib/easy_lock.h
---- spack-src/lib/easy_lock.h	2022-06-26 13:58:05.000000000 -0700
-+++ spack-src-patched/lib/easy_lock.h	2022-08-05 07:59:16.450137961 -0700
+From e2e7f54b7bea521fa8373095d0f43261a720cda0 Mon Sep 17 00:00:00 2001
+From: Daniel Stenberg <daniel@haxx.se>
+Date: Mon, 27 Jun 2022 08:46:21 +0200
+Subject: [PATCH] easy_lock.h: include sched.h if available to fix build
+
+Patched-by: Harry Sintonen
+
+Closes #9054
+---
+ lib/easy_lock.h | 3 +++
+ 1 file changed, 3 insertions(+)
+
+diff --git a/lib/easy_lock.h b/lib/easy_lock.h
+index 819f50ce815b8..1f54289ceb2d3 100644
+--- a/lib/easy_lock.h
++++ b/lib/easy_lock.h
 @@ -36,6 +36,9 @@
  
  #elif defined (HAVE_ATOMIC)

--- a/var/spack/repos/builtin/packages/curl/package.py
+++ b/var/spack/repos/builtin/packages/curl/package.py
@@ -120,6 +120,9 @@ class Curl(AutotoolsPackage):
     # curl queries pkgconfig for openssl compilation flags
     depends_on("pkgconfig", type="build")
 
+    # https://github.com/curl/curl/pull/9054
+    patch("easy-lock-sched-header.patch", when="@7.84.0")
+
     @classmethod
     def determine_version(cls, exe):
         curl = Executable(exe)

